core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java

```diff
@@ -584,6 +606,26 @@ public Builder localNodeId(String localNodeId) {
             return this;
         }
 
+        public DiscoveryNode nodeWithSameAddress(DiscoveryNode node) {
```

Comment:
It is only called when a node that joins the cluster is added on the master node. In test code it is also asserted as an invariant on `DiscoveryNodes`. As the number of nodes is limited I prefer keeping it out of the `nodes` map for now.